src/NoteDataUtil.cpp

```diff
@@ -1855,6 +1855,144 @@ static void GetTrackMapping( StepsType st, NoteDataUtil::TrackMapping tt, int Nu
 	}
 }
 
+static void HyperShuffleNotes( NoteData &inout, int iStartIndex, int iEndIndex)
+{
+	int iNumTracks = inout.GetNumTracks();
+
+	// Stores the list of tracks that can have notes placed into them.
+	// This will be shuffled so that notes can be placed in different tracks.
+	vector<int> viTargetTracks;
+	// Stores the list of tap notes that need to be placed.
+	vector<TapNote> vtnTargetTaps;
+	viTargetTracks.reserve(iNumTracks);
+	vtnTargetTaps.reserve(iNumTracks);
+
+	vector<int> viHoldEndRows;
+	viHoldEndRows.resize(iNumTracks);
+	fill(viHoldEndRows.begin(), viHoldEndRows.end(), -1);
+	
+	// A "free track" is a track that is not part of a hold and is therefore
+	// free to be shuffled into.
+	int iFreeTracks = iNumTracks;
+
+	// If this is the first row, there cannot be any active holds yet.
+	if ( iStartIndex != 0 )
+	{
+		// Search for any hold notes that might be present at the first row.
+		// Once the loop below gets going, it keeps track of hold notes itself.
+		for ( int track = 0; track < iNumTracks; track++ )
+		{
+			int iTargetRow = iStartIndex;
+
+			if ( inout.GetPrevTapNoteRowForTrack(track, iTargetRow) ) 
+			{
+				const TapNote &tn = inout.GetTapNote(track, iTargetRow);
+
+				// Check for a hold that ends on or after this row.
+				if ( tn.type == TapNoteType_HoldHead )
+				{
+					int iHoldEndRow = iTargetRow + tn.iDuration;
+
+					if ( iHoldEndRow >= iStartIndex )
+					{
+						viHoldEndRows[track] = iHoldEndRow;
+						iFreeTracks--;
+					}
+				}	
+			}
+		}
+	}
+
+	FOREACH_NONEMPTY_ROW_ALL_TRACKS_RANGE( inout, r, iStartIndex, iEndIndex )
+	{
+		viTargetTracks.clear();
+		vtnTargetTaps.clear();
+		int iActualTaps = 0;
+
+		for( int track=0; track<iNumTracks; track++ )
+		{
+			int iHoldEndRow = viHoldEndRows[track];
+
+			// Check if this track is actually still occupied by a hold
+			if(iHoldEndRow != -1)
+			{
+				if (r > iHoldEndRow)
+				{
+					iHoldEndRow = -1;
+					viHoldEndRows[track] = iHoldEndRow;
+					iFreeTracks++;
+				}
+			}
+
+			const TapNote tn1 = inout.GetTapNote(track, r);
+			switch( tn1.type )
+			{
+			case TapNoteType_HoldTail:
+				ASSERT_M(0, ssprintf("saw a HoldTail during HyperShuffleNotes (row %d, track %d)", r, track));
+
+				//if asserts are off, delete it and treat it like an empty note
+				//(if they are on, this code is not reachable)
+				inout.SetTapNote(track, r, TAP_EMPTY);
+
+				[[fallthrough]];
```

Comment:
Huh I didn't even know we compiled ITGmania with C++17

Reply:
Yeah I can take that out. I assumed it wouldn't be an issue because C++11 has attribute support even if that attribute doesn't exist.

Reply:
Probably best to take it out for now, and just replace it with a comment: `// Fallthrough`